### PR TITLE
Add Statistic plugin (#1581 Amount of hours spent on a book)

### PR DESCRIPTION
**Pull request by apletnev**
New plugin Statistic. Provides information about dates and time spent on read books. 
It saves data in to directory statistic files like this:
```javascript
return {
	["total_time"] = 29467,
	["authors"] = "Николай А. Островский",
	["title"] = "Как закалялась сталь",
	["highlights"] = 0,
	["notes"] = 0,
	["details"] = {
		[1] = {
			["read"] = 8,
			["time"] = 1425168796,
			["page"] = 3
		},
		[2] = {
			["read"] = 63,
			["time"] = 1425168872,
			["page"] = 9
		},
		[3] = {
			["read"] = 39,
			["time"] = 1425168911,
			["page"] = 10
		}
         } 
	["series"] = "",
	["language"] = "ru"
}
```
Based on this information creates reports: total time for all books, and time spent for current book. 

Looks like

Current:
current

Total
total

Total by book
totaldetails

### CODE REVIEW

File: frontend/apps/reader/modules/readerhighlight.lua [hunk 1]
```diff
@@ -384,9 +384,11 @@ function ReaderHighlight:highlightFromHoldPos()
     end
 end
 
+--[[
```
Comment (NiLuJe) [inline frontend/apps/reader/modules/readerhighlight.lua]: Why has that been commented? (it happens to break the testsuite, so something relies on it).

File: plugins/statistic.koplugin/main.lua [hunk 2]
```diff
@@ -0,0 +1,423 @@
...
+        self.data = self:importFromFile(book_properties.title .. ".stat")
+        self:savePropertiesInToData(book_properties)
+        self.data.pages = self.view.document:getPageCount()
+        return
+    end
+end
+
+function ReaderStatistic:addToMainMenu(tab_item_table)
+    table.insert(tab_item_table.plugins, {
+        text = _("Statistic"),
```
Comment (NiLuJe) [inline plugins/statistic.koplugin/main.lua]: Should be "Statistics" w/ an s ;).

File: plugins/statistic.koplugin/main.lua [hunk 3]
```diff
@@ -0,0 +1,423 @@
...
+    }
+
+    book_status.close_callback = function()
+        UIManager:close(self.status_menu)
+    end
+
+    book_status.show_parent = self.status_menu
+
+    return {
+        text = "Current",
```
Comment (NiLuJe) [inline plugins/statistic.koplugin/main.lua]: Untranslated?

File: plugins/statistic.koplugin/main.lua [hunk 4]
```diff
@@ -0,0 +1,423 @@
...
+    }
+
+    self.total_status.close_callback = function()
+        UIManager:close(self.total_menu)
+    end
+
+    self.total_status.show_parent = self.total_menu
+
+    return {
+        text = "Total",
```
Comment (NiLuJe) [inline plugins/statistic.koplugin/main.lua]: TR?

File: plugins/statistic.koplugin/main.lua [hunk 5]
```diff
@@ -0,0 +1,423 @@
...
+end
+
+function ReaderStatistic:exportToFile(book_properties)
+    if book_properties then
+        self:savePropertiesInToData(book_properties)
+    end
+
+    local statistics = io.open(statistic_dir .. "/" .. self.data.title .. ".stat", "w")
+    if statistics then
+        local current_locale = os.setlocale()
```
Comment (NiLuJe) [inline plugins/statistic.koplugin/main.lua]: I get the intent here, I just hope this won't be an issue on slightly broken embedded systems ;).
Not sure you have a better way to deal with this though, so this is just just a random comment ;).

**Comment by NiLuJe** [comment]
Looks cool :).
Made a few comments nitpicking about grammar, sorry ;).

File: plugins/statistic.koplugin/main.lua (hunk 2 above)
Comment (apletnev) [inline plugins/statistic.koplugin/main.lua]: Ok, will change.

File: frontend/apps/reader/modules/readerhighlight.lua (hunk 1 above)
Comment (chrox) [inline frontend/apps/reader/modules/readerhighlight.lua]: Yes, Travis CI failed exactly here:
```
35mError[0m â†’ [36mspec/front/unit/readerhighlight_spec.lua[0m @ [36m150[0m
[1mReaderhighlight module highlight for PDF documents for reflowed page should response on tap gesture[0m
spec/front/unit/readerhighlight_spec.lua:39: attempt to call method 'onHighlight' (a nil value)

stack traceback:
	spec/front/unit/readerhighlight_spec.lua:39: in function 'tap_highlight_text'
	spec/front/unit/readerhighlight_spec.lua:151: in function <spec/front/unit/readerhighlight_spec.lua:150>
```
Please fix this.

**Comment by chrox** [comment]
Awesome plugin. :+1:

**Comment by chrox** [comment]
BTW, the build log for openssl is so verbose that Travis cannot show the final result. Maybe we need to silent openssl build at koreader-base.

File: plugins/statistic.koplugin/main.lua (hunk 4 above)
Comment (apletnev) [inline plugins/statistic.koplugin/main.lua]: What do you mean?

File: plugins/statistic.koplugin/main.lua (hunk 3 above)
Comment (apletnev) [inline plugins/statistic.koplugin/main.lua]: For add translation i have to execute "make pot"?

Comment (Frenzie) [inline plugins/statistic.koplugin/main.lua]: No, encapsulate in _()

Cheers

On September 9, 2015 8:32:25 PM GMT+02:00, Alexander Pletnev <notifications@github.com> wrote:
>> +        cface = Font:getFace("cfont", 20),
>> +    }
>> +
>> +    self.status_menu = CenterContainer:new {
>> +        dimen = Screen:getSize(),
>> +        book_status,
>> +    }
>> +
>> +    book_status.close_callback = function()
>> +        UIManager:close(self.status_menu)
>> +    end
>> +
>> +    book_status.show_parent = self.status_menu
>> +
>> +    return {
>> +        text = "Current",
>
>For add translation i have to execute "make pot"? 
>
>---
>Reply to this email directly or view it on GitHub:
>https://github.com/koreader/koreader/pull/1628/files#r39079192

-- 
Sent from my Android device with K-9 Mail. Please excuse my brevity.

File: plugins/statistic.koplugin/main.lua (hunk 4 above)
Comment (houqp) [inline plugins/statistic.koplugin/main.lua]: He meas this text is not translatable, need to wrap it with `_()`

**Comment by apletnev** [comment]
Guys, i updated the CR. Please review it. 
Travis shows error, looks like it is not dependent on my code.

**Comment by apletnev** [comment]
Updated, please check. Thanks.

File: plugins/statistics.koplugin/main.lua [hunk 6]
```diff
@@ -0,0 +1,424 @@
...
+local Screen = require("device").screen
+local DEBUG = require("dbg")
+local Menu = require("ui/widget/menu")
+local Font = require("ui/font")
+local _ = require("gettext")
+local TimeVal = require("ui/timeval")
+local dump = require("dump")
+local lfs = require("libs/libkoreader-lfs")
+local tableutil = require("tableutil")
+local T = require("ffi/util").template
```
Comment (Frenzie) [inline plugins/statistics.koplugin/main.lua]: This is total nitpicking, but I do believe there is a touch of organization to the requires, with the main stuff generally ending with

```lua
local DEBUG = require("dbg")
local T = require("ffi/util").template
local _ = require("gettext")
```

But perhaps I'm imagining that. If I am I would still like it to be the case. :P

**Comment by Frenzie** [comment]
Anyway, it seems good to me. Nice work!

File: plugins/statistics.koplugin/main.lua [hunk 7]
```diff
@@ -0,0 +1,424 @@
...
+                read = dates[date_text].read + v.read,
+                count = dates[date_text].count + 1,
+                date = dates[date_text].date
+            }
+        end
+    end
+
+    table.insert(result, { text = _(book.title) })
+    for k, v in tableutil.spairs(dates, function(t, a, b) return t[b].date > t[a].date end) do
+        table.insert(result, { text = _(k), mandatory = T(_("Pages (%1) Time: %2"), v.count, os.date("!%X", v.read)) })
```
Comment (Frenzie) [inline plugins/statistics.koplugin/main.lua]: Btw, the _(k) doesn't really seem to serve any purpose?

Comment (houqp) [inline plugins/statistics.koplugin/main.lua]: Yeah, we don't need _() for k here.

Comment (apletnev) [inline plugins/statistics.koplugin/main.lua]: Fixed, see new commit 85f284440c19ead8f4de1b2fab5190dbc09819ba

**Comment by Frenzie** [comment]
I figure it's good to go; does everyone agree? Maybe consider squashing the commits together with e.g. `git rebase -i HEAD~4`. :)

**Comment by chrox** [comment]
Cool, merging now.

**Pull request merged**